Rpm cleanup

akirapop opened the pull request:
This code is ready to merged into master. The cron job on the ci is scheduled to run daily at 26:00 (02:00 am).

akirapop committed:
First commits.

akirapop committed:
Bug fix.

akirapop committed:
Restructured script to run partly on local, partly on dh.

akirapop committed:
Bug fixing

akirapop committed:
Still bugfixing.

akirapop committed:
Still bug-fixing...

akirapop committed:
Still bugchecking.

akirapop committed:
Still trying to find source of git error.

akirapop committed:
Still trying...

akirapop committed:
Still bug-fixing...

akirapop committed:
Still...

akirapop committed:
Still trying to find out what's going on.

akirapop committed:
Trying again.

akirapop committed:
Trying more things.

akirapop committed:
One more try

akirapop committed:
Bug fixed ... I hope!

akirapop committed:
Cleanup stale remote refs.

akirapop committed:
Added code to cleanup rpms in master/

akirapop committed:
Silly bracket error fixed.

akirapop committed:
Fixed silly > vs. -gt error.

akirapop committed:
Verifying rm code.

akirapop committed:
Final touches.

akirapop committed:
Final touches.

akirapop committed:
Almost there.

akirapop committed:
Almost there.

akirapop committed:
Strangeness...

akirapop committed:
Strangeness...

akirapop committed:
Damnded quotes

akirapop committed:
Things seem ok now.

akirapop committed:
Cleaning things up a little.

akirapop committed:
Testing new Jenkins configuration for this branch.

akirapop committed:
Testing new Jenkins configuration for this branch.

akirapop committed:
Added new Jenkinsfile for cleaning up rpm repos.

akirapop committed:
Renamed Jenkins.trash_remvoal --> Jenkins.garbage_collection to reflect jenkins job name

akirapop committed:
Cleaning up the Jenkins.garbage_collection file.

akirapop committed:
Testing git checkout.

akirapop committed:
Made final (?) edits to script prior to pull request.

unakatsuo commented on Jenkinsfile:
```diff
@@ -104,7 +114,8 @@ if( buildParams.BUILD_OS != "all" ){
+//stage_rpmbuild(label)
+//stage_integration(label)
+  stage_clean_rpms(label)
 }
```
Please revert the changes to ``Jenkinsfile``. This causes the existing build process halts and cleanup tasks is performed by another Jenkins job not by ``citest`` job.

akirapop replied:
Return the Jenkinsfile to its orginal, is that what you mean? Jenkinsfileは、元に戻すっていうこと、でしょう？

unakatsuo replied:
Why did you change this file? ``Jenkinsfile.garbage_collection``, isn't it the script file to perform cleanup job?

akirapop replied:
Do you mean, why did I move the file? I just thought it made more sense to reside in the ci/repo_cleanups directory, since that is what it is logically connected to. The jenkins configuration was updated to reflect this new location.

unakatsuo replied:
I mean why this branch has the change to the root ``Jenkinsfile``.  I understand that why ``Jenkinsfile.garbage_collection`` is added and there is new Jenkins job ``garbage_collection`` refers to ``Jenkinsfile.garbage_collection``.

I estimate that you applied changes to the root ``Jenkinsfile`` at the beginning and later copied the modified ``Jenkinsfile`` to ``Jenkinsfile.garbage_collection`` at af29f96eb2b24. But your change still remain on the root ``Jenkinsfile``.

https://github.com/axsh/openvdc/tree/d6c8cea3e7fc742733c3cb4dba61bd6d272e2e64

akirapop replied:
I commented out the stage_xxx steps because they are unrelated to what this branch does. I think it is a waste of resources to have Jenkins run builds and tests on code not relevant to this branch.

unakatsuo replied:
So, this branch is going to be merged. What will happen?

I don't complain about the series of changes to the file during development meantime. Why are the changes still appearing at the HEAD of this branch?

Metallion replied:
I agree with @unakatsuo on this one.

It's true that the main Jenkinsfile runs tests generally unrelated to this branch but if it gets merged with those lines commented out, it will stop running the tests on master (and future branches) too. It's ok to comment these out during development but before merge, it should be restored back to how it was.

Also in general I think it's good to run tests at least once before merging even if unrelated because you just never know. Some times unrelated code can have unexpected consequences.

akirapop replied:
OK, I've replaced this Jenkinsfile with the current version in master.

unakatsuo commented on Jenkinsfile.garbage_collection:
```diff
@@ -0,0 +1,85 @@
+    // https://issues.jenkins-ci.org/browse/JENKINS-26133
+    RELEASE_SUFFIX=sh(returnStdout: true, script: "./deployment/packagebuild/gen-dev-build-tag.sh").trim()
+}
+
```
Please cleanup the code. I think the lines of code by the line here has no meaning for cleanup job.

akirapop replied:
Jenkinsfile.garbage_collection now has only code directly related to the repo. cleanup

Metallion replied:
As mentioned above, you seem to have forgotten to remove this file when you moved it to `ci/repo_cleanups`.

You can use `git rm` followed by `git commit` to remove it.

unakatsuo commented on repo_cleanups/remove_stale_rpm.sh:
```diff
@@ -0,0 +1,231 @@
+#!/bin/bash 
```
This file should be moved under ``ci/``.

akirapop replied:
Ok, I've moved repo_cleanups/ to this directory.

Metallion replied:
It looks like you forgot to call `git rm` here too. For the future, `git mv` is a useful command to rename files with git being aware of it. `git mv foo bar` is basically the same as:

```
mv foo bar
git add bar
git rm foo
```

unakatsuo reviewed (changes requested):
I found the bunch of  lines which should be removed. Please fix.

akirapop committed:
Responded to rpm cleanup requests.

akirapop committed:
Bug fix.

akirapop committed:
More fixing.

akirapop committed:
Still fixing...

Metallion commented on Jenkinsfile.garbage_collection:
```diff
@@ -0,0 +1,85 @@
+BRANCH=${env.BRANCH_NAME}
+"""
+  writeFile(file: "build.env", text: build_env)
+}
```
~~I haven't read the actual cleanup script yet but this looks like code copy-pasted from the main Jenkinsfile. I think there are many variables here that are not used. For example there is `LEAVE_CONTAINER` but I don't think this job starts any containers.~~

~~To keep the code clean and readable, any unused variables should be removed.~~

I see below that a copy of this file was committed with the code cleaned up. I guess you just forgot to call `git rm` on this.

akirapop replied:
As commented below, the file has been removed.

Metallion commented on ci/repo_cleanups/Jenkinsfile.garbage_collection:
```diff
@@ -0,0 +1,26 @@
+def checkout_and_merge() {
+    checkout scm
+    sh "git merge origin/master"
+}
```
This function is not used (and isn't needed in the garbage collection job) so it should be removed.

akirapop replied:
It has been removed.

akirapop replied:
Likewise, Jenkinsfile.garbage_collection has been removed from the top of the tree for this branch.

Metallion commented on ci/repo_cleanups/Jenkinsfile.garbage_collection:
```diff
@@ -0,0 +1,26 @@
+
+node() {
+    stage "Checkout"
+//  checkout_and_merge()
```
I think it's better to remove commented out code before merging. Git makes it possible to get old code back and I think it makes the code cleaner.

akirapop replied:
I have removed this comment.

Metallion commented on ci/repo_cleanups/Jenkinsfile.garbage_collection:
```diff
@@ -0,0 +1,26 @@
+//  checkout_and_merge()
+    stage_clean_rpms()
+}
+
```
Other than those two comments, it's a clean file now. ^_^ Very good.

Metallion reviewed (changes requested):
One more general comment I have is the commit messages.  They should be a little more descriptive of the change you made.

Here's a couple PRs with examples of good commit messages.

https://github.com/axsh/openvdc/pull/68
https://github.com/axsh/openvdc/pull/58

No need to go back and change commit messages but just something to keep in mind for the future. ;)

akirapop committed:
The Jenkinsfile from the master branch has been reinstated. In addition, some small cleanup
has been done on ci/repo_cleanups/Jenkinsfile.garbage_collection.

Metallion commented on ci/repo_cleanups/remove_stale_rpm.sh:
```diff
@@ -0,0 +1,231 @@
+script=$(write_script)
+
+## Send the script generated above to the dh machine
+echo -e "${script/_STUB_/${git_branches}/}" | $SSH_REMOTE /usr/bin/bash 
```
Writing the whole script in a function is not a good idea I think. For starters it removes syntax highlighting and it's likely to have other side effects too.

If it is required to send the script to another machine to execute it there, it's better to save the script normally as its own file and then send that through SSH either directly from the Jenkinsfile or from a wrapper script.

Metallion commented on ci/repo_cleanups/remove_stale_rpm.sh:
```diff
@@ -0,0 +1,231 @@
+
+
+function run_git {
+#   git for-each-ref --sort=-committerdate refs/remotes --format='%(HEAD)%(refname:short), %(committerdate:relative)'
```
As mentioned above I think it's cleaner to remove old code. They're still preserved in git history if it's needed to bring them back later.

Metallion commented on ci/repo_cleanups/remove_stale_rpm.sh:
```diff
@@ -0,0 +1,231 @@
+        fi
+
+        if [[ ${ndays} -gt 14 ]]; then
+            ## Just in case the continue command does not work as I think it should...
```
What is the concern you have with the continue command? The way it works is documented well in the [bash man page](https://linux.die.net/man/1/bash).

> continue [n]
Resume the next iteration of the enclosing for, while, until, or select loop. If n is specified, resume at the nth enclosing loop. n must be ≥ 1. If n is greater than the number of enclosing loops, the last enclosing loop (the ''top-level'' loop) is resumed. When continue is executed inside of loop, the return value is non-zero when n is ≤ 0; Otherwise, continue returns 0 value. When continue is executed outside of loop, the return value is 0.

I also used to write code like this but I think it's better to spend a little extra time and understand exactly what the commands you use are doing. That way there's no need for clauses like this and it simplifies the code.

Metallion commented on ci/repo_cleanups/remove_stale_rpm.sh:
```diff
@@ -0,0 +1,231 @@
+
+    var=$(cat <<"EOF"
+
+    declare -A  git_branches    ## Store the list of git branches in a hash
```
I wonder if there is a way to do it without using hashes. They require bash version 4 and it's a complicated data structure which is not what bash is generally good at.

I haven't understood the script enough yet to provide a good alternative but I'll keep reading a bit more.

Metallion commented on ci/repo_cleanups/remove_stale_rpm.sh:
```diff
@@ -0,0 +1,231 @@
+               echo "Cannot remove \"${no_rm}\". Ignoring."
+               return 0
+           fi
+        done
```
Still haven't understood the script completely but relying on what the current working directory is seems a bit of a red flag to me. I would prefer constructing the absolute path to be deleted. We already have the `rpm_repo_dir` variable so it shouldn't be too hard to do.

I plan to read some more to see what exactly is being done and then I'll give more concrete suggestions. :)

Metallion commented on ci/repo_cleanups/remove_stale_rpm.sh:
```diff
@@ -0,0 +1,231 @@
+        rm -rf ./${dead_dir}
+        if [ $? -ne 0 ]; then
+           echo "       rm FAILED for ./${dead_dir} "
+        fi
```
Instead of error checking, it's better to put `set -xe` at the top of your script.

The `-e` flag will cause your script to fail and stop execution if any of its subprocesses (such as `rm`) exit with a status other than 0.

The `-x` flag will print the lines currently executed to stdout.

Metallion reviewed (changes requested):
Still reviewing the script but I already placed a couple comments that I wanted to publish. :)

akirapop committed:
Did lots of refactoring here, primarily pullling out code used to assess which subdirectories
are "stale". Each directory is named after a git branch name. The refactored code determines
which directories no longer have a corresponding git branch -or- have a git branch which is
has not been pushed to within 2 weeks. This code can/will be used for the cache cleaning code
as well.

akirapop committed:
Made a few necessary adjustments to this file. This includes specifying the required node (.111) to run this on

akirapop committed:
Fixed small bug. But a larger one lurks somewhere, still.

akirapop committed:
Running with -x flag to find this bug.

akirapop committed:
Made a change to the git for-each-ref command.
The git version on the ci machines (virutal and bare metal) is
a little old (1.8.3) and so the date formats are not so flexible.
The best I could see to do was to have the dates printed in the
form yyyy-mm-dd and remove each '-' with sed.

akirapop committed:
Fixed an errant/missing parens.

akirapop committed:
More debugging. (This must be done inside Jenkins, alas.)

akirapop committed:
More debugging.

akirapop committed:
More checking.

Metallion commented on ci/garbage_collection/repo_cleanups/remove_stale_rpm.sh:
```diff
@@ -0,0 +1,229 @@
+#!/bin/bash 
```
This file still has the old code and isn't called anywhere. I guess you forgot to run `git rm`?

Metallion commented on Jenkinsfile:
```diff
@@ -38,16 +38,18 @@ RELEASE_SUFFIX=${RELEASE_SUFFIX}
 GIT_BRANCH=${env.BRANCH_NAME}
```
The root Jenkinsfile should be made the same as the current master branch. It seems you have copied over the Jenkinsfile from your local master branch but that one was not up-to-date with the master branch on github. Here's how I suggest we fix it.

In the openvdc root directory:

```
# Bring your local master up-to-date
git checkout master
git pull --rebase

# Merge updated local master into your branch. This will cause a merge conflict on Jenkinsfile.
git checkout rpm_cleanup
git merge master

# Get the Jenkinsfile from current master on github.
curl https://raw.githubusercontent.com/axsh/openvdc/master/Jenkinsfile -o Jenkinfile

# Solve merge conflict.
git add Jenkinsfile
git commit
```

Metallion reviewed (changes requested):
I see this output in the CI:

```
Checking for stale rpm repos under master...
rm -rf /var/www/html/openvdc-repos/20170110092208git66d1ba5  (command not executed)
rm -rf /var/www/html/openvdc-repos/20170111053806git384446e  (command not executed)
rm -rf /var/www/html/openvdc-repos/20170111072135gite7faf4c  (command not executed)
```

It checks for old commits under `/var/www/html/openvdc-repos/` but the master repos are under `/var/www/html/openvdc-repos/master/`.

akirapop committed:
Made one bugfix (added master to appropriate dir names) and removed an unused file.

akirapop committed:
Merge branch 'master' into rpm_cleanup

akirapop commented:
I have made the following changes to the code:  I have factored out the portions of the code that determine which directories (which are mapped to git branch names) need to be removed: Does the git branch no longer exist? yes - Remove. If "no" - has the branch been pushed to within 14 days? No? Remove. This code can now be incorporated also into the cache_cleaning code. 

I've changed the directory structure:
     openvdc/ci/repo_cleanups/garbage_collection_misc.sh
                                              /prune_branches.sh
                                              /repo_cleanups/
                                                                       /rpm_garbage_collection.sh

<on the cache_cleanup branch>
                                              /cache_cleanups/
                                                                        /cache_garbage_collection.sh

akirapop commented:
Other changes:

   (1) The script runs directly on the dh now, so there is no writing the script to a sh variable and then
         running that via ssh

   (2) No hash is used to find directories that need to be pruned. Instead, the comm command
        (file comparison command) is used.

akirapop committed:
Added line to allow for direct execution -- the code is now live, so it *will* now
remove directories! I also made a bug-fix: my code to check against removing "master"
and "develop" directories was flawed. I have fixed that code now and tested it locally.
